Updates for ITKv5.2 code cleanups ... ROUND1

hjmjohnson opened the pull request.

hjmjohnson committed:
COMP: Fix ambiguous type conversion for vnl_matrix

hjmjohnson committed:
STYLE: Fixed compiler warnings and logical bugs.

hjmjohnson committed:
STYLE: Prefer = default to explicitly trivial implementations

This check replaces default bodies of special member functions with
= default;. The explicitly defaulted function declarations enable more
opportunities in optimization, because the compiler might treat
explicitly defaulted functions as trivial.

Additionally, the C++11 use of = default more clearly expreses the
intent for the special member functions.

hjmjohnson committed:
STYLE: Remove redundant void argument lists

Find and remove redundant void argument lists.

hjmjohnson committed:
STYLE: Use default member initialization

Converts a default constructor’s member initializers into the new
default member initializers in C++11. Other member initializers that match the
default member initializer are removed. This can reduce repeated code or allow
use of ‘= default’.

hjmjohnson committed:
STYLE: Use default member initialization

Converts a default constructor’s member initializers into the new
default member initializers in C++11. Other member initializers that match the
default member initializer are removed. This can reduce repeated code or allow
use of ‘= default’.

hjmjohnson committed:
STYLE: Prefer = default to explicitly trivial implementations

This check replaces default bodies of special member functions with
= default;. The explicitly defaulted function declarations enable more
opportunities in optimization, because the compiler might treat
explicitly defaulted functions as trivial.

Additionally, the C++11 use of = default more clearly expreses the
intent for the special member functions.

hjmjohnson committed:
DOC: Update copyright assignment to NumFOCUS

The mission of NumFOCUS is to promote open
practices in research, data, and scientific
computing.

https://numfocus.org

phcerdan commented on include/itkRieszRotationMatrix.hxx:
```diff
@@ -209,7 +208,7 @@ RieszRotationMatrix< T, VImageDimension >
   if ( this->m_Order == 1 )
     {
-    S = this->GetSpatialRotationMatrix().GetVnlMatrix();
+    S = this->GetSpatialRotationMatrix().GetVnlMatrix().as_matrix();
```
What's the difference, why is `as_matrix()` needed?

hjmjohnson replied:
This is for type matching.  GetVnlMatrix() returns a vnl_fixed_matrix<r,c>, but InternalMatrixType is vnl_matrix.  The old implicit (and somewhat dangerous) conversions are now disabled and require explicit matching.

phcerdan commented on test/itkWaveletCoeffsPhaseAnalyzisImageFilterTest.cxx:
```diff
@@ -126,53 +124,92 @@ itkWaveletCoeffsPhaseAnalyzisImageFilterTest( int argc, char *argv[] )
-    }
+  std::cerr << " failed!" << std::endl;
+  std::cerr << waveletFunction << " wavelet type not supported." << std::endl;
+  return EXIT_FAILURE;
```
Why the duplication of the code for each wavelet is needed?

hjmjohnson replied:
Not my code.  You would need to contact the original author.

phcerdan replied:
In your commit you are duplicating (4x) a chunk of code that @jhlegarreta or myself authored.

I am not asking about the original chunk of code, but about the reasons of duplication.

Maybe this is correct to avoid a shadowing of `WaveletType`, but just asking for your insight, if any, or maybe this is the result of an automatic refactor.

phcerdan reviewed (commented):
Looks good, a couple of questions below.
Thanks @hjmjohnson

mseng10 committed:
ENH: Update enums from changes in ITK

hjmjohnson commented on test/itkWaveletCoeffsSpatialDomainImageFilterTest.cxx:
```diff
@@ -101,64 +101,96 @@ itkWaveletCoeffsSpatialDomainImageFilterTest( int argc, char *argv[] )
   using ImageType = itk::Image<float, ImageDimension>;
 
   using WaveletScalarType = double;
-  using WaveletType = itk::SimoncelliIsotropicWavelet< WaveletScalarType, ImageDimension>;
```
@phcerdan Oh.  I forgot I changed this.  

The original code was wrong.  'using' directives are compile-time settings, but the original code was treating it as a runtime variable.  The original code only compiled WaveletCoeffsSpatialDomainImageFilterType with 'WaveletType = itk::SimoncelliIsotropicWavelet'.  All the  other 'using' statementments in the if/else sections were "Shadow types that are unused"

phcerdan replied:
I see, thanks @hjmjohnson !

jhlegarreta commented:
:100: **thanks** for doing this Hans.

The CI failure is due to the **NumFOCUS** _vs_. **Insight** header mismatch. Updating the `ITKGitTag` in the Azure pipelines YAML to the latest ITK version should fix the issue.

hjmjohnson committed:
ENH: Add .gitattributes to allow running ITK clang-formatting scripts

```
git filter-branch -f \
--tree-filter "~/ITK/Utilities/Maintenance/clang-format.bash --clang-format ~/Dashboard/src/ITK-clang11/clang-format-Linux --tracked" \
    master..
```

dzenanz commented:
Pablo, merge when satisfied! The latest force push should take care of NumFOCUS issue.
